README.md

```diff
@@ -2,7 +2,217 @@
+A typical example is the addition of a `private var` in a `trait`, e.g., going from
+
+```scala
+trait A {
+  var x: Int = 1
+  def foo(): Int = x
+}
+```
+
+to
+
+```scala
+trait A {
+  var x: Int = 1
+  private var y: Int = 2
+  def foo(): Int = x + y
+}
+```
+
+When mixing `A` into a class `C`, the compiler needs to create a field for `y` in `C`.
+If `C` was compiled against the former version of `A`, but is linked against the latter, linking breaks.
+On the JVM, this materializes as various forms of `LinkageError` at run-time.
+In Scala.js and Scala Native, it materializes as linking errors at link time, i.e., at the time of producing a `.js` file or an executable.
+
+MiMa detects that this change is invalid, and reports it so that it can be fixed before shipping erroneous versions of a library.
+
+### Consequences of TASTy incompatibility
+
+Similarly to binaries, changes in a Scala program can be TASTy incompatible even though we think they are compatible.
+For example, adding a `super` call within the body of a trait method, or switching between a `Seq` and a vararg parameter.
+
+Breaking TASTy compatibility in the API of a library `L` can cause issues when relinking the library with another library `M` that depends on it, when used from a project `P` that depends on both.
+In particular, problems can arise if `M` has `inline def`s that call into the API of `L`.
+
+Here is a concrete example.
+
+```scala
+// L.scala
+object L {
+  def list(xs: Seq[Int]): List[Int] = xs.toList
+}
+```
+
+```scala
+// M.scala
+object M {
+  inline def foo(): String =
+    L.list(Seq(1, 2, 3)).sum.toString
+}
+```
+
+```scala
+// Test.Scala
+object Test {
+  def main(args: Array[String]): Unit = {
+    println(M.foo())
+  }
+}
+```
+
+We first compile everything, and we even verify that it runs:
+
+```
+$ cs launch scalac:3.1.2 -- -d bin/ L.scala M.scala Test.scala
+$ cs launch scala:3.1.2 -- -cp bin/ Test
+6
+```
+
+Then, we change `xs: Seq[Int]` into `xs: Int*` in `L.scala`:
+
+```scala
+// L.scala
+object L {
+  def list(xs: Int*): List[Int] = xs.toList
+}
+```
+
+and recompile only `L`:
+
+```
+$ cs launch scalac:3.1.2 -- -d bin L.scala
+```
+
+This change is not detected by MiMa, as it is binary compatible.
+
+Finally we try to recompile only `Test`:
+
+```
+$ cs launch scalac:3.1.2 -- -cp bin/ -d bin/ Test.scala
+-- [E007] Type Mismatch Error: Test.scala:3:17 ---------------------------------
+3 |    println(M.foo())
+  |            ^^^^^^^
+  |            Found:    Seq[Int]
+  |            Required: Int
+  | This location contains code that was inlined from M.scala:3
+
+longer explanation available when compiling with `-explain`
+1 error found
+```
+
+This happens even though, from the point of view of `Test` the change of `L` is both source and binary compatible, because it is not TASTy-compatible.
```

Comment:
It could be helpful to detail a little bit what is causing the incompatibility, and why `inline` plays an important role here.